src/adc/v2.rs

```diff
@@ -0,0 +1,265 @@
+use embassy_hal_internal::into_ref;
+
+use super::blocking_delay_us;
+use crate::adc::{Adc, AdcChannel, Instance, Resolution, SampleTime};
+use crate::pac::adc::vals::Extsel;
+use crate::peripherals::ADC;
+// use crate::time::Hertz;
+use crate::{rcc, Peripheral};
+use crate::pac::RCC;
+
+// mod ringbuffered_v2;
+// pub use ringbuffered_v2::{RingBufferedAdc, Sequence};
+
+/// Default VREF voltage used for sample conversion to millivolts.
+pub const VREF_DEFAULT_MV: u32 = 3300;
+/// VREF voltage used for factory calibration of VREFINTCAL register.
+pub const VREF_CALIB_MV: u32 = 3300;
+
+pub struct VrefInt;
+impl AdcChannel<ADC> for VrefInt {}
+impl super::SealedAdcChannel<ADC> for VrefInt {
+    fn channel(&self) -> u8 {
+        17
+    }
+}
+
+impl VrefInt {
+    /// Time needed for internal voltage reference to stabilize
+    pub fn start_time_us() -> u32 {
+        10
+    }
+}
+
+pub struct Temperature;
+impl AdcChannel<ADC> for Temperature {}
+impl super::SealedAdcChannel<ADC> for Temperature {
+    fn channel(&self) -> u8 {
+        16
+    }
+}
+
+impl Temperature {
+    /// Time needed for temperature sensor readings to stabilize
+    pub fn start_time_us() -> u32 {
+        10
+    }
+}
+
+// pub struct Vbat;
+// impl AdcChannel<ADC> for Vbat {}
+// impl super::SealedAdcChannel<ADC> for Vbat {
+//     fn channel(&self) -> u8 {
+//         18
+//     }
+// }
+
+pub enum Prescaler {
+    Div2,
+    Div4,
+    Div6,
+    Div8,
+}
+
+impl Prescaler {
+    // I couldn't find the MAX_FREQUENCY for the PY32. 
+    // Therefore, the division factor is left for the user to decide.
```

Comment:
In general, the maximum frequency of the ADC module is not recommended to exceed 16MHz

Reply:
Thanks